assets/js/blocks/cart-checkout/subscriptions-integration.js

```diff
@@ -0,0 +1,56 @@
+/**
+ * External dependencies
+ */
+import {
+	ExperimentalOrderShippingPackages,
+	ShippingRatesControlPackage,
+} from '@woocommerce/blocks-checkout';
+import { registerPlugin } from '@wordpress/plugins';
+import { useStoreCart, useSelectShippingRate } from '@woocommerce/base-hooks';
+import { useMemo } from '@wordpress/element';
+
+const RenderSubscriptionPackages = () => {
+	const {
+		extensions: { subscriptions },
+	} = useStoreCart();
+
+	// Flatten all packages from recurring carts.
+	const packages = useMemo( () => {
+		const newPackages = [];
+
+		Object.values( subscriptions ).forEach( ( recurringCart ) => {
+			const recurringCartPackages = recurringCart.shipping_rates || [];
+
+			recurringCartPackages.forEach( ( recurringCartPackage ) => {
+				newPackages.push( recurringCartPackage );
+			} );
+		} );
+
+		return newPackages;
+	}, [ subscriptions ] );
+
+	const { selectShippingRate, selectedShippingRates } = useSelectShippingRate(
+		packages
+	);
+
+	return packages.map( ( { package_id: packageId, ...packageData } ) => (
+		<ExperimentalOrderShippingPackages key={ packageId }>
+			<SubscriptionPackage
+				key={ packageId }
+				packageData={ packageData }
+				onSelectRate={ ( newShippingRate ) => {
+					selectShippingRate( newShippingRate, packageId );
+				} }
+				selected={ selectedShippingRates[ packageId ] }
```

Comment:
Nothing hererelies on external data and all can be performed internally, given that we're publishing a very opinioned shipping package selector, would it make sense to abstract this part?

Reply:
Ya I concur with Nadir here. It looks like the only subscriptions specific thing here is getting subscriptions from the extensions property and converting that to a packages array. Could we just have extensions register their own callback for modifying the packages when additional shipping areas are created?  So in this case in whatever interface is provided Subscriptions would just provide this as a callback for packages:

```js
const getPackages = ( cart ) => {
	const { subscriptions } = cart.extensions;
	return Object.values( subscriptions )
		.map( ( recurringCart ) => recurringCart.shipping_rates )
		.filter( Boolean )
		.flat();
};
```

Notice the callback would receive the cart object so the extension could use that to derive its packages from (like what you are doing currently in this code).

Reply:
Is there any prior art for this suggestion?

Reply:
> Is there any prior art for this suggestion?

For my comment or for Darren's comment?

>  Could we just have extensions register their own callback for modifying the packages when additional shipping areas are created?

This part is not obvious to me, do you mean having a callback on `ShippingRatesControlPackage` like this:

```js
const SubscriptionPackage = ( props ) => {
	const getPackages = useCallback( ( cart ) => {
		const { subscriptions } = cart.extensions;
		return Object.values( subscriptions )
			.map( ( recurringCart ) => recurringCart.shipping_rates )
			.filter( Boolean )
			.flat();
	}, [] );
	return (
		<ShippingRatesControlPackage { ...props } packageData={ getPackages } />
	);
};
```

Reply:
I think we can combine this feedback with https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/3687#discussion_r562164364